x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/persistence/TrainedModelCacheMetadataService.java

```diff
@@ -123,9 +122,11 @@ public ClusterState execute(BatchExecutionContext<CacheMetadataUpdateTask> batch
                 return initialState;
             }
 
-            return ClusterState.builder(initialState)
-                .metadata(Metadata.builder(initialState.metadata()).putCustom(TrainedModelCacheMetadata.NAME, currentCacheMetadata))
-                .build();
+            final var finalMetadata = currentCacheMetadata;
+            return initialState.copyAndUpdateProject(
+                initialState.metadata().getProject().id(),
+                builder -> builder.putCustom(TrainedModelCacheMetadata.NAME, finalMetadata)
+            );
```

Comment:
Can we keep the existing structure for this one as well. It gets me a bit that we have to re-assign the variable for lambda.